Fix/gravityforms amp plus

adekbadek opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Closes #1371.

Some changes were made to Jetpack Instant Search code (#1486) – @miguelpeixe please take a look to see if I haven't messed it up.

### How to test the changes in this Pull Request:

1. While on `master`, install [Gravity Forms plugin](https://www.gravityforms.com/)
2. Create a form* with a single text input, visit and AMP page and verify the AMP-originating "It appears your submission was successful…" message is displayed on submission (#1371)
3. Switch to this branch, verify the AMP-originating message is not displayed after submission and instead, GF's "Thank you for contacting us…" message is displayed
4. Create a form where a field uses conditional logic:

<img width="1243" alt="image" src="https://user-images.githubusercontent.com/7383192/156390599-8a8708a4-b427-45f7-b23b-7b0fb728d3e5.png">

5. Load the page, observe that the form with conditional logic is **not displayed**
6. Install & activate GF's Polls plugin and add a poll form to the page
7. Observe that after submitting the choice in the poll, the results are unstyled
8. Enable AMP Plus (`wp config set NEWSPACK_AMP_PLUS_ENABLED true --raw`)
9. Enable AMP Plus' handling of Gravity Form (`wp config set NEWSPACK_AMP_PLUS_GRAVITYFORMS true --raw`)
10. Load the page, observe 
    1. both simple and conditional forms are loading and can be submitted 
    2. the poll form displays results with styling

\* Visit `/wp-admin/admin.php?page=gf_new_form`, create the form, then insert it in a page via GF's block

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201860761661613
  - https://app.asana.com/0/0/1201528767060186

dkoo commented on includes/plugins/class-gravityforms.php:
```diff
@@ -0,0 +1,466 @@
+			return $args;
+		}
+		$args['ajax']           = false;
+		$args['enableHoneypot'] = '0';
```
Why do we need to disable the honeypot field? My understanding is that this is a pretty widely used feature to prevent spam responses.

adekbadek replied:
Yeah, I've reverted this change (aff2645c25023ab0b1dfa3c547bf9b4a0833ef24) after testing with a honeypot-enabled form

dkoo replied:
Confirmed that a poll and a form with honeypot fields still captured entries correctly.

dkoo reviewed (changes requested):
A lot of publishers will love this! A few issues I ran into:

1. For polls, if you click the "View results" button, sometimes but not always you'll see a duplicate "Back to the poll" button like so:

<img width="808" alt="Screen Shot 2022-03-02 at 12 57 41 PM" src="https://user-images.githubusercontent.com/2230142/156441282-0740a3ec-744a-4468-9c01-78386c297261.png">

2. When clicking "Back to the poll", it triggers a JS error: 

```
Uncaught Error: No ampdoc found for javascript:void(0)
```

3. When submitting a form or poll, the non-AMP behavior hides the form when displaying the confirmation message. In contrast, the AMP Plus behavior does not hide the form, making it easy to spam responses. It's especially awkward for polls because the results are shown but you can also click "View results" on the unhidden poll and see two sets of the results:

<img width="812" alt="Screen Shot 2022-03-02 at 1 10 52 PM" src="https://user-images.githubusercontent.com/2230142/156442102-c455d816-417b-4146-bd6f-75d3ff0d7197.png">

<img width="817" alt="Screen Shot 2022-03-02 at 1 11 21 PM" src="https://user-images.githubusercontent.com/2230142/156441820-c0396bb0-5be0-40f6-9529-c30e2dc6c0ed.png">

4. More of a suggestion than an issue, but the messaging (both success and error) could use a bit of styling. Currently they use the default text styling and therefore bleed in with the rest of the post content:

<img width="831" alt="Screen Shot 2022-03-02 at 1 01 25 PM" src="https://user-images.githubusercontent.com/2230142/156442048-14d2316a-9bd3-4f09-aa74-28a9b00cccb0.png">

<img width="444" alt="Screen Shot 2022-03-02 at 1 00 46 PM" src="https://user-images.githubusercontent.com/2230142/156442044-3b466f99-289c-4aa2-bff6-952468a8c4da.png">

5. For polls, if using the "Enable 'other' choice" setting, it shows a text input for readers to input their own text. However, this text input is uneditable in AMP Plus mode.

6. I'm not sure this is related to this PR per se, but Jetpack Instant Search seems partially broken. I couldn't get it to work on every page load, but it does work sometimes. When it doesn't, it always results in loading the search results template page but with no results found, even when the search term would return results in both Instant Search and a regular WP search. I'm not able to pinpoint exact conditions for replicating it—it seems somewhat random if I refresh the site or open it in a new tab.

adekbadek commented on includes/plugins/class-gravityforms.php:
```diff
@@ -0,0 +1,466 @@
+
+		if ( isset( $form['enableHoneypot'] ) && '1' === $form['enableHoneypot'] ) {
+			// Remove randomly generated honeypot fields.
+			array_pop( $form['fields'] );
```
(honeypot-disabling reverted in aff2645c25023ab0b1dfa3c547bf9b4a0833ef24)

adekbadek commented on includes/plugins/class-gravityforms.php:
```diff
@@ -0,0 +1,466 @@
+
+		wp_dequeue_script( 'jquery' );
+
+		$error_message        = __( 'There is a mistake in the form!', 'newspack' );
```
These are fallback messages – by default, GF's messages will be used. Added filtering in eab266102064636bc7e4fb890019d21d9596ee90

adekbadek commented on includes/plugins/class-gravityforms.php:
```diff
@@ -0,0 +1,466 @@
+	 * @param string $content The post content.
+	 */
+	public static function handle_gf_polls_blocks( $content ) {
+		if ( ! self::should_disable_scripts() && has_block( 'gravityforms/polls' ) ) {
```
Thanks! Added in fc62f42658fa0e1f1401f2f6e884654e6b1e45e9

dkoo replied:
Confirmed that it now also works with a poll inside the Gravity Forms block.

adekbadek commented:
1.  

> For polls, if you click the "View results" button, sometimes but not always you'll see a duplicate "Back to the poll" button like so

I was not able to reproduce that, were you able to pinpoint any conditions?

2. 

> When clicking "Back to the poll", it triggers a JS error:

I see it too, but it does not seem to cause any issues?

3. 

> When submitting a form or poll, the non-AMP behavior hides the form when displaying the confirmation message. 

Fixed in 8c7c2d2f79b4f4ca895d152b48aebdb65f43b772 – the submitted form should disappear now.

4.  

> More of a suggestion than an issue, but the messaging (both success and error) could use a bit of styling. Currently they use the default text styling and therefore bleed in with the rest of the post content:

That's something to think about, but out of scope of this PR IMO. This PR only aims to enable use of GF plugin with AMP Plus.

5. 

> For polls, if using the "Enable 'other' choice" setting, it shows a text input for readers to input their own text. However, this text input is uneditable in AMP Plus mode.

Fixed in 47f92ef1f99806b98d48b9ff483c00a4ec0ef623, thanks for catching that!

6. 

> I'm not sure this is related to this PR per se, but Jetpack Instant Search seems partially broken. […]

Were you able to confirm that the issue is not appearing on `master` branch?

dkoo reviewed (changes requested):
Looking much better after the last round! 

> I was not able to reproduce that, were you able to pinpoint any conditions?

I wasn't able to replicate this after the latest changes, so perhaps we were missing a script or style somewhere.

> > I'm not sure this is related to this PR per se, but Jetpack Instant Search seems partially broken. […]

> Were you able to confirm that the issue is not appearing on master branch?

I'm not able to replicate on McMansion and I'm having trouble getting Instant Search to work on my new local environment, so let's defer this until it comes up again.

Just a couple of lingering issues:

1. For both forms and polls, after submitting (or clicking "view results" on the poll) the "submit" buttons are still visible and interactive, so it's still possible to spam submissions:

<img width="591" alt="Screen Shot 2022-04-05 at 1 31 37 PM" src="https://user-images.githubusercontent.com/2230142/161836551-47463e21-c80c-4119-b1f8-7427b1a1b372.png">

<img width="385" alt="Screen Shot 2022-04-05 at 1 31 23 PM" src="https://user-images.githubusercontent.com/2230142/161836652-7f892955-69e9-471a-8b57-6291b63fc3cd.png">


2. For polls, after submitting your answer, the "view results" button is still visible, so you can still see two polls if you click it:
<img width="394" alt="Screen Shot 2022-04-05 at 1 31 29 PM" src="https://user-images.githubusercontent.com/2230142/161836630-247dd53c-740d-4ef5-8529-55e6dde1d973.png">

adekbadek commented:
Fixed in ec950f4bfc6769925d856be02b8613523f81575c, thanks for catching that!

dkoo reviewed (approved):
Latest change did the trick for me! 👍
